File: C++/SelectionSort/src/SelectionSort.cpp
#include "SelectionSort.h"

void SelectionSort::byAuthor(vector<Book>& _books) {

	for(int i = 0; i < _books.size(); i++) {
		int minIndex = i;
		for(int j = i+1; j < _books.size(); j++) {

			if(_books[j].compareAuthor(_books[minIndex]) < 0) {
				minIndex = j;
			}
		}
		if(minIndex != i) {
			swapBooks(_books, i, minIndex);
		}
	}
}

void SelectionSort::byNumberPages(vector<Book>& _books) {

	for(int i = 0; i < _books.size(); i++) {
		int minIndex = i;
		for(int j = i+1; j < _books.size(); j++) {

			if(_books[j].compareNumberPages(_books[minIndex]) < 0) {
				minIndex = j;
			}
		}
		if(minIndex != i) {
			swapBooks(_books, i, minIndex);
		}
	}
}

void SelectionSort::byTitle(vector<Book>& _books) {

	for(int i = 0; i < _books.size(); i++) {
		int minIndex = i;
		for(int j = i+1; j < _books.size(); j++) {

			if(_books[j].compareTitle(_books[minIndex]) < 0) {
				minIndex = j;
			}
		}
		if(minIndex != i) {
			swapBooks(_books, i, minIndex);
		}
	}
}

void SelectionSort::swapBooks(vector<Book>& _books, int _swap1, int _swap2) {
	Book temp = _books[_swap1];
	_books[_swap1] = _books[_swap2];
	_books[_swap2] = temp;
}

File: JavaScript/SelectionSort/src/SelectionSort.js

  function SelectionSort(){};

  SelectionSort.byAuthor = function(_books) {
    for(var i = 0; i < _books.length; i++) {
      var minIndex = i;
      for(var j = i+1; j < _books.length; j++) {
        if(_books[j].compareAuthor(_books[minIndex]) < 0) {
          minIndex = j;
        }
      }
      if (minIndex != i) {
        SelectionSort.swapBooks(_books, i, minIndex);
      }
    }
  }

  SelectionSort.byNumberPages = function(_books) {
    for(var i = 0; i < _books.length; i++) {
      var minIndex = i;
      for(var j = i+1; j < _books.length; j++) {
        if(_books[j].compareNumberPages(_books[minIndex]) < 0) {
          minIndex = j;
        }
      }
      if (minIndex != i) {
        SelectionSort.swapBooks(_books, i, minIndex);
      }
    }
  }

  SelectionSort.byTitle = function(_books) {
    for(var i = 0; i < _books.length; i++) {
      var minIndex = i;
      for(var j = i+1; j < _books.length; j++) {
        if(_books[j].compareTitle(_books[minIndex]) < 0) {
          minIndex = j;
        }
      }
      if (minIndex != i) {
        SelectionSort.swapBooks(_books, i, minIndex);
      }
    }
  }

  SelectionSort.swapBooks = function(_books, _swap1, _swap2) {
    var temp = _books[_swap1];
    _books[_swap1] = _books[_swap2];
    _books[_swap2] = temp;
  }

  module.exports = SelectionSort;
